odoo/addons/base/models/ir_mail_server.py

```diff
@@ -17,13 +18,21 @@
 from socket import gaierror, timeout
 from OpenSSL import crypto as SSLCrypto
 from OpenSSL.crypto import Error as SSLCryptoError, FILETYPE_PEM
-from OpenSSL.SSL import Error as SSLError
-from urllib3.contrib.pyopenssl import PyOpenSSLContext
+from OpenSSL.SSL import Error as SSLError, VERIFY_PEER, VERIFY_FAIL_IF_NO_PEER_CERT
+from urllib3.contrib.pyopenssl import PyOpenSSLContext, get_subj_alt_name
 
 from odoo import api, fields, models, tools, _
 from odoo.exceptions import UserError
 from odoo.tools import ustr, pycompat, formataddr, email_normalize, encapsulate_email, email_domain_extract, email_domain_normalize
 
+try:
+    from urllib3.util.ssl_match_hostname import CertificateError, match_hostname
```

Comment:
urllib3 versions we care about are:
* 1.24 (debian buster, python3.7, odoo15)
* 1.25 (ubuntu focal)
* 1.26 (debian bullseye/bookworm, ubuntu jammy)
* 2.0 (ubuntu noble)

The function originated in Python 3.3 as `ssl.match_hostname` but has been deprecated in Python 3.7. It supports only domain names and not ip addresses.

For ages the urllib3 team did vendor the function, I have been able to retrace its inclusion to 2011. The function was vendored as-is until ~2016. Inside a version older than 1.24 the urllib team started to take ownership on that function and added a bunch of fixes/patches, e.g. to support IP Address inside Subject Alternative Name (x509 extension). So we wanna use *their* `match_hostname` for the added features.

In 1.24/1.25 the function should be imported from `urllib3.packages.ssl_match_hostname`. They moved it to `urllib3.util.ssl_match_hostname` in 1.26.

That last fallback on the python's function is actually useless because urllib already is smart about it:

https://github.com/urllib3/urllib3/blob/9b95f29c575d73260ec4f5a8c0ea368cf242019e/src/urllib3/packages/ssl_match_hostname/__init__.py#L3-L16

Reply:
I did add the missing comments